src/static/js/Changeset.js

```diff
@@ -903,20 +907,24 @@ class TextLinesMutator {
     let removed = '';
     if (this._isCurLineInSplice()) {
```

Comment:
What if `this._curCol` is greater than or equal to the current line's length? In that case, shouldn't we set `this._curCol` to 0 and increment `this._curLine` before this check? Or is that case not possible due to invariants?

If invariants make that case impossible, I think it would be worthwhile to add an assert statement.

Reply:
skip:
- if L present, curCol is set to 0 in all branches (in skipLines)
- if L not present, curCol will be +=N (number of skipped chars) [0]

insert:
- if L present, curCol is set to 0 after applying L, because of the "multiline end with newline" constraint
- if L is present, but curLine was not in spliceXXX, curCol isn't touched
- if L is not present and we are processing the last line of lines array, curCol is increased by N chars [1]
- if L is not present and we are not processing the last line, curCol is increased by N chars [2]

remove:
- no changes to curCol

In cases [1] and [2] curCol calculation is safe, as a text string is given to the respective function (the term that's added to curCol is derived from from text).

Also after calling `close` or skipping over multiple L that don't have attributes, the splice is not active anymore and it's possible that curCol is > 0. So there are four possible "start" states:
- splice not active, curCol is 0 [X1]
- splice not active, curCol > 0 [X2]
- splice active, curCol is 0 [X3]
- splice active, curCol > 0 [X4]

So for all 4 "start" states, I'll add coverage for [0] exceeding the number of chars in the line (ie a text of "test\n" and skipping over 6 chars so we end in the middle of the next line), followed by a call to removeLines.

If we want to assert I think it's better to do it where curCol is changed.